### PR TITLE
Support for Swift Package Manager

### PR DESCRIPTION
**Pull request by sja26**
Continuation of #122 

I replaced ZipArchive with the SPM compatible ZIPFoundation, at the cost of increasing the minimum deployment target to macOS 10.11 and iOS 9.0.

I've added a Makefile for running the Swift Package tests. I'll look into adding this to the Travis CI setup, and checking out Linux support.

I'd also draw your attention to 122dc08, I'm not sure what differed from the previous Clibgit2 package that caused the `malloc: Incorrect checksum for freed object crash`?

### CODE REVIEW

**Comment by mdiep** [comment]
Awesome! Thanks. I'll try to review this soon.

File: SwiftGit2/Repository.swift [hunk 1]
```diff
@@ -124,6 +124,8 @@ final public class Repository {
 	///
 	/// Returns a `Result` with a `Repository` or an error.
 	class public func at(_ url: URL) -> Result<Repository, NSError> {
+		SwiftGit2Init
```
Comment (tengyifei) [inline SwiftGit2/Repository.swift]: Slight nit-picking: While this does work, my opinion is that while injecting side-effects into global variable initializer is reasonable, but the way it's used looks a bit unnatural. Upon first glance one could not easily see that this is reading from the `SwiftGit2Init` variable, thereby executing `git_libgit2_init` exactly once.

How about:

```swift
// In Libgit2.swift, or SwiftGit2.swift
private var gitInitialized: Int = ({
	return Int(git_libgit2_init())
})()

func ensureGitInitialized() {
	precondition(gitInitialized >= 0)
}

// In places we want to use git, such as Repository.swift
ensureGitInitialized()
// ... start using git
```

File: SwiftGit2Tests/RepositorySpec.swift [hunk 2]
```diff
@@ -23,7 +24,7 @@ class RepositorySpec: QuickSpec {
 				let url = URL(fileURLWithPath: "blah")
 				let result = Repository.at(url)
 				expect(result.error?.domain) == libGit2ErrorDomain
-				expect(result.error?.localizedDescription).to(match("Failed to resolve path"))
+				expect(result.error?.localizedDescription.lowercased()).to(beginWith("failed to resolve path"))
```
Comment (tengyifei) [inline SwiftGit2Tests/RepositorySpec.swift]: Perhaps something like

```swift
expect(result.error?.localizedDescription).to(match("(F|f)ailed to resolve path.*"))
```

would be a bit more tolerating to different libgit2 versions?

File: SwiftGit2/SwiftGit2.swift [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+//  SwiftGit2
+//
+//  Created by Andrew Breckenridge on 10/23/17.
+//  MIT.
+//
+
+import Clibgit2
+
+// swiftlint:disable:next identifier_name
+let SwiftGit2Init: () = {
```
Comment (tengyifei) [inline SwiftGit2/SwiftGit2.swift]: Suggest turning this to a private variable and wrapping with a function. Please refer to my previous comment.

File: SwiftGit2.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -909,6 +917,10 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = BEB31F4E1A0D75EE00F525B9 /* iOS-Application.xcconfig */;
 			buildSettings = {
+				HEADER_SEARCH_PATHS = (
+					"$(OBJROOT)/UninstalledProducts/include",
```
Comment (tengyifei) [inline SwiftGit2.xcodeproj/project.pbxproj]: What is `UninstalledProducts`?

**Comment by ios3** [comment]
What's the wait on this PR?

**Comment by satishbabariya** [comment]
@sja26 @mdiep This Swift Packege https://github.com/fmccann/Clibgit2 will require "libgit2 to installed in System" as it uses SPM's System Library Targets.
Ref: https://github.com/apple/swift-package-manager/blob/master/Documentation/PackageDescriptionV4_2.md

**Comment by tonyarnold** [comment]
Has this succumbed to the "Unfinished SPM support" curse that's plaguing this repo?

File: SwiftGit2/Repository.swift (hunk 1 above)
Comment (sja26) [reply SwiftGit2/Repository.swift]: Addressed in commit https://github.com/SwiftGit2/SwiftGit2/pull/138/commits/5a28aea51a610f3e2222bfe0724e76bbd8018b07

File: SwiftGit2/SwiftGit2.swift (hunk 3 above)
Comment (sja26) [reply SwiftGit2/SwiftGit2.swift]: Addressed in commit https://github.com/SwiftGit2/SwiftGit2/pull/138/commits/5a28aea51a610f3e2222bfe0724e76bbd8018b07

File: SwiftGit2.xcodeproj/project.pbxproj (hunk 4 above)
Comment (sja26) [reply SwiftGit2.xcodeproj/project.pbxproj]: I don't know I guess Xcode added it when I was changing the Swift version in the .xcodeproj file?

File: SwiftGit2Tests/RepositorySpec.swift (hunk 2 above)
Comment (sja26) [reply SwiftGit2Tests/RepositorySpec.swift]: Addressed in commit  https://github.com/SwiftGit2/SwiftGit2/pull/138/commits/43fa649e00ebe4b458b184103ad681af96b4a7dd

File: Package.swift [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+		.package(url: "https://github.com/Quick/Quick", from: "1.3.2"),
+		.package(url: "https://github.com/Quick/Nimble", from: "7.3.1"),
+		.package(url: "https://github.com/weichsel/ZIPFoundation", .upToNextMajor(from: "0.9.0"))
+    ],
+    targets: [
+        .target(
+            name: "SwiftGit2",
+            dependencies: ["Result"],
+	    	path: "SwiftGit2"),
+		.testTarget(
```
[the remark this answers is not in the document]
Comment (sja26) [reply Package.swift]: Addressed in commit  https://github.com/SwiftGit2/SwiftGit2/pull/138/commits/d0b24feeda9b19bd3ea9edd8cea6b0c784832428

File: script/cibuild [hunk 6]
```diff
@@ -83,4 +84,10 @@ elif [ "$SCHEME" == "SwiftGit2-iOS" ]; then
     ${XCODE_OPTIONS[*]} \
     build test \
     2>&1 | xcpretty $XCPRETTY_FORMAT_OPTIONS
+elif [ "$SCHEME" == "SwiftGit2-SPM" ]; then
+  echo "*** Building and testing $SCHEME..."
+  echo
+
+  make test \
+    2>&1 | xcpretty $XCPRETTY_FORMAT_OPTIONS
```
[the remark this answers is not in the document]
Comment (sja26) [reply script/cibuild]: Addressed in commit  https://github.com/SwiftGit2/SwiftGit2/pull/138/commits/98dccdaf011dfccc64f377b30ad0611c9aaab89b

File: .travis.yml [hunk 7]
```diff
@@ -9,6 +9,7 @@ env:
   matrix:
     - SCHEME="SwiftGit2-OSX"
     - SCHEME="SwiftGit2-iOS"
+    - SCHEME="SwiftGit2-SPM"
```
[the remark this answers is not in the document]
Comment (sja26) [reply .travis.yml]: I guess I did kind of shoehorn it in there 😬

File: Package.swift [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+		.package(url: "https://github.com/Quick/Quick", from: "1.3.2"),
+		.package(url: "https://github.com/Quick/Nimble", from: "7.3.1"),
+		.package(url: "https://github.com/weichsel/ZIPFoundation", .upToNextMajor(from: "0.9.0"))
+    ],
+    targets: [
+        .target(
+            name: "SwiftGit2",
+            dependencies: ["Result"],
+			path: "SwiftGit2"),
+		.testTarget(
```
Comment (mdiep) [inline Package.swift]: This indentation is still off

File: .travis.yml [hunk 9]
```diff
@@ -17,10 +17,14 @@ before_install:
...
   - gem install xcpretty-travis-formatter
+  - brew update && brew bundle
 
 install: script/bootstrap
 
-script: script/cibuild
+script:
+  - script/cibuild
+  - make build 2>&1 | xcpretty -f `xcpretty-travis-formatter`
+  - make test 2>&1 | xcpretty -f `xcpretty-travis-formatter`
```
Comment (mdiep) [inline .travis.yml]: Can you separate the SwiftPM build to a different TravisCI job?

**Comment by cameroncooke** [comment]
@sja26 @mdiep Is this dead? Really need SwiftPM support for macOS at least and this seemed so close to getting over the line?

**Comment by sja26** [comment]
@cameroncooke SwiftPM support lives on in PR https://github.com/SwiftGit2/SwiftGit2/pull/159🤞

**Comment by cameroncooke** [comment]
Seeing a 404 when trying to view this PR?

**Comment by genkobar** [comment]
> Seeing a 404 when trying to view this PR?
> […](#)
> On 17 Mar 2020, at 20:59, sja26 ***@***.***> wrote: ﻿ @cameroncooke SwiftPM support lives on in PR #159🤞 — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

@cameroncooke It seems GitHub decided to include the emoji hand in the link - try this: https://github.com/SwiftGit2/SwiftGit2/pull/159

**Pull request closed without merging**